src/fusiontensor/array_cast.jl

```diff
@@ -0,0 +1,193 @@
+# This file defines interface to cast from and to generic array
+
+using BlockArrays: AbstractBlockArray, BlockedArray, blockedrange, blocklengths, findblock
+
+using BlockSparseArrays: BlockSparseArrays, BlockSparseArray
+using GradedUnitRanges: AbstractGradedUnitRange, blocklabels
+using SymmetrySectors: block_dimensions, quantum_dimension
+using TensorAlgebra: contract
+
+# =================================  High level interface  =================================
+
+#### cast from array to symmetric
+function FusionTensor(
+  array::AbstractArray,
+  codomain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+  domain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+)
+  return cast_from_array(array, codomain_legs, domain_legs)
+end
+
+#### cast from symmetric to array
+function BlockSparseArrays.BlockSparseArray(ft::FusionTensor)
+  return cast_to_array(ft)
+end
+
+# =================================  Low level interface  ==================================
+function cast_from_array(
+  array::AbstractArray,
+  codomain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+  domain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+)
+  bounds = block_dimensions.((codomain_legs..., domain_legs...))
+  blockarray = BlockedArray(array, bounds...)
+  return cast_from_array(blockarray, codomain_legs, domain_legs)
+end
+
+function cast_from_array(
+  blockarray::AbstractBlockArray,
+  codomain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+  domain_legs::Tuple{Vararg{AbstractGradedUnitRange}};
+  tol::Float64=1e-12,
+)
+  # input validation
+  if length(codomain_legs) + length(domain_legs) != ndims(blockarray)  # compile time
+    throw(DomainError("legs are incompatible with array ndims"))
+  end
+  if quantum_dimension.((codomain_legs..., domain_legs...)) != size(blockarray)
+    throw(DomainError("legs dimensions are incompatible with array"))
+  end
+
+  ft = unsafe_cast_from_array(blockarray, codomain_legs, domain_legs)
+
+  # if blockarray is not G-invariant, norm(ft) < norm(blockarray)
+  if abs(norm(ft) - norm(blockarray)) > tol
+    throw(
+      InexactError(
+        :FusionTensor, typeof(blockarray), typeof(codomain_legs), typeof(domain_legs)
+      ),
+    )
+  end
+  return ft
+end
+
+function unsafe_cast_from_array(
+  blockarray::AbstractBlockArray,
+  codomain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+  domain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+)
+  ft = FusionTensor(eltype(blockarray), codomain_legs, domain_legs)
+  for (f1, f2) in keys(trees_block_mapping(ft))
+    b = findblock(ft, f1, f2)
+    ft[f1, f2] = contract_fusion_trees(blockarray[b], f1, f2)
+  end
+  return ft
+end
+
+function cast_to_array(ft::FusionTensor)
```

Comment:
I think we can just call this `to_array(...)`.

Reply:
If there's no ambiguity, could you also just write `Base.Array(::FusionTensor)`?

Reply:
I think the issue with that is that it actually outputs a `BlockSparseArray`. We could call it `BlockSparseArrays.BlockSparseArray`. But I also like the idea of having a generic converter from `FusionTensor` to some kind of non-fused array, which is agnostic from the user perspective about what the non-fused format will be. So maybe:
```julia
function BlockSparseArrays.BlockSparseArray(ft::FusionTensor)
  # This function implementation we are discussing
end

# Maybe come up with a better name
unfuse(ft::FusionTensor) = BlockSparseArray(ft)
```

Reply:
Also, it seems like we should support:
```julia
function BlockSparseArrays.BlockSparseArray(ft::FusionTensor{<:Any,N}, axes::Tuple{Vararg{AbstractUnitRange,N}}) where {N}
  # ...
end
```
so that you can attach graded axes to the output, say to specify a set of abelian symmetries. Right now I see it just makes it blocked but doesn't have any sectors on it which makes sense since as we've discussed it isn't necessarily well defined what abelian group the original symmetry group will reduce to (though if it is abelian to start with we could just use the same abelian group by default).

Is there a way to choose a reasonable guess for an abelian group given a non-abelian group?